RT-1.29 BGP chained policies README (#2595)
openconfig/featureprofiles: feature/bgp/policybase/otg_tests/chained_policies/README.md

# RT-1.29: BGP chained import/export policy attachment

## Summary

- A list of policies to be attached to a neighbor's import-policy
- A list of policies to be attached to a neighbor's export-policy
- Applicable to both IPv4 and IPv6 BGP neighbors

## Testbed type

*   https://github.com/openconfig/featureprofiles/blob/main/topologies/atedut_2.testbed

## Procedure

> 

Reviewer: In the future we may change the style guide to require this.  This more accurately represents the way we apply configuration and sometimes reveals issues with DUT behavior.

```suggestion
### Applying configuration

For each section of configuration below, prepare a gnmi.SetBatch  with all the configuration items appended to one SetBatch.  Then apply the configuration to the DUT in one gnmi.Set using the `replace` option.

```

    *   /routing-policy/policy-definitions/policy-definition/statements/statement/conditions/match-prefix-set/config/prefix-set

##### Configure another route-policy to set the local preference
*   Configure an IPv4 route-policy definition with the name ```lp-policy-v4```
    *   /routing-policy/policy-definitions/policy-definition/config/name
*   For routing-policy ```lp-policy-v4``` configure a statement with the name ```lp-statement-v4```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
*   For routing-policy ```lp-policy-v4``` statement ```lp-statement-v4``` set policy-result as ```ACCEPT_ROUTE```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
##### Configure BGP actions to set local-pref
*   For routing-policy ```lp-policy-v4``` statement ```lp-statement-v4``` set local-preference to ```200```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-local-pref

##### Configure chained bgp import policies for the DUT BGP neighbor on ATE Port-1

> *   Set default import policy to ```REJECT_ROUTE```

Reviewer: ```suggestion
*   Set default import policy to ```REJECT_ROUTE``` (Note: even though this is the OC default, the DUT should still accept this configuration)
```

    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy

> *   Apply both the policies as a chain/list ```[match-policy-v4, lp-policy-v4]```

Reviewer: ```suggestion
*   Add both policies in the specified order to the leaf-list `import-policy`:   ```[match-policy-v4, lp-policy-v4]```
```

    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/import-policy

##### Verification

> *   Verify that chained import policies are successfully applied to the DUT BGP neighbor on ATE Port-1

Reviewer: ```suggestion
*   Use gNMI `replace` to send the configuration to the DUT.
*   Use gNMI `subscribe` with mode `once` to retrieve the configuration `state` from the DUT.  This is to confirm the chained import policies are successfully applied to the DUT BGP neighbor on ATE Port-1
```

##### Configure another route-policy to set the MED
*   Configure an IPv4 route-policy definition with the name ```med-policy-v4```
    *   /routing-policy/policy-definitions/policy-definition/config/name
*   For routing-policy ```med-policy-v4``` configure a statement with the name ```med-statement-v4```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
*   For routing-policy ```med-policy-v4``` statement ```med-statement-v4``` set policy-result as ```ACCEPT_ROUTE```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
##### Configure BGP actions to set MED
*   For routing-policy ```med-policy-v4``` statement ```med-statement-v4``` set MED to ```1000```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med

##### Configure chained bgp export policies for the DUT BGP neighbor on ATE Port-1
*   Set default export policy to ```REJECT_ROUTE```
    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy

> *   Apply both the policies as a chain/list ```[asp-policy-v4, med-policy-v4]```

Reviewer: ```suggestion
*   Add both policies in order to the `export-policy` leaf-list, ie: ```[asp-policy-v4, med-policy-v4]```
```

##### Configure another route-policy to set the local preference
*   Configure an IPv6 route-policy definition with the name ```lp-policy-v6```
    *   /routing-policy/policy-definitions/policy-definition/config/name
*   For routing-policy ```lp-policy-v6``` configure a statement with the name ```lp-statement-v6```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
*   For routing-policy ```lp-policy-v6``` statement ```lp-statement-v6``` set policy-result as ```ACCEPT_ROUTE```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
##### Configure BGP actions to set local-pref
*   For routing-policy ```lp-policy-v6``` statement ```lp-statement-v6``` set local-preference to ```200```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-local-pref

##### Configure chained bgp import policies for the DUT BGP neighbor on ATE Port-1
*   Set default import policy to ```REJECT_ROUTE```
    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy

> *   Apply both the policies as a chain/list ```[match-policy-v6, lp-policy-v6]```

Reviewer: ```suggestion
*   Add both policies in order to the leaf-list `import-policy`: ```[match-policy-v6, lp-policy-v6]```
```

##### Configure another route-policy to set the MED
*   Configure an IPv6 route-policy definition with the name ```med-policy-v6```
    *   /routing-policy/policy-definitions/policy-definition/config/name
*   For routing-policy ```med-policy-v6``` configure a statement with the name ```med-statement-v6```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
*   For routing-policy ```med-policy-v6``` statement ```med-statement-v6``` set policy-result as ```ACCEPT_ROUTE```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
##### Configure BGP actions to set MED
*   For routing-policy ```med-policy-v6``` statement ```med-statement-v6``` set MED to ```1000```
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med

##### Configure chained bgp export policies for the DUT BGP neighbor on ATE Port-1
*   Set default export policy to ```REJECT_ROUTE```
    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy

> *   Apply both the policies as a chain/list ```[asp-policy-v6, med-policy-v6]```

Reviewer: ```suggestion

*   Add both policies in order to the leaf-list `export-policy`:  ```[asp-policy-v6, med-policy-v6]```
```

*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/export-policy

## Telemetry parameter coverage

*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/import-policy
*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/export-policy
*   /network-instances/network-instance/protocols/protocol/bgp/rib/attr-sets/attr-set/as-path/as-segment/state/member
*   /network-instances/network-instance/protocols/protocol/bgp/rib/afi-safis/afi-safi/ipv6-unicast/loc-rib/routes/route/prefix
*   /network-instances/network-instance/protocols/protocol/bgp/rib/afi-safis/afi-safi/ipv4-unicast/loc-rib/routes/route/prefix
*   /network-instances/network-instance/protocols/protocol/bgp/rib/attr-sets/attr-set/state/med

## Protocol/RPC Parameter Coverage

* gNMI

>   * Get

Reviewer: ```suggestion
  * Subscribe (ONCE)
```

>   * Set

Reviewer: ```suggestion
  * Set (REPLACE)
```
